### PR TITLE
executor: fix logging format of prepared statements (#16062)

### LINKED ISSUE #16062: The slow log of `Prepare` statements don't record the real SQL
## Description
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
The slow log of `Prepare` statements don't record the real SQL . And it's not similar with MySQL .

```sql

MySQL [test]> PREPARE mystmt FROM 'SELECT sleep(?)';
Query OK, 0 rows affected (0.01 sec)

MySQL [test]> SET @number = 5;
Query OK, 0 rows affected (0.00 sec)

MySQL [test]> EXECUTE mystmt USING @number;
+----------+
| sleep(?) |
+----------+
|        0 |
+----------+
1 row in set (5.00 sec)

MySQL [test]> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                 |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v2.1.17
Git Commit Hash: 603f0b95d391c336132f33de6877feb6e6984e54
Git Branch: HEAD
UTC Build Time: 2020-04-02 10:31:39
GoVersion: go version go1.13.6 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

<!-- If possible, provide a recipe for reproducing the error. -->

### 2. What did you expect to see?
**In Mysql:**
```sql
mysql> PREPARE mystmt FROM 'SELECT sleep(?)';
Query OK, 0 rows affected (0.00 sec)
Statement prepared

mysql> SET @number = 5;
Query OK, 0 rows affected (0.00 sec)

mysql> EXECUTE mystmt USING @number;
+----------+
| sleep(?) |
+----------+
|        0 |
+----------+
1 row in set (5.00 sec)

mysql> \! tail -n 10 /home/mysql/mysqllog/mysql-slow.log
# Time: 2020-04-02T22:32:17.643358-05:00
# User@Host: root[root] @  [127.0.0.1]  Id:  1572
# Query_time: 5.000391  Lock_time: 0.000000 Rows_sent: 1  Rows_examined: 0
SET timestamp=1585881137;
SELECT sleep(5);
# Time: 2020-04-03T06:15:05.959133-05:00
# User@Host: root[root] @  [172.16.4.142]  Id:  1573
# Query_time: 5.000401  Lock_time: 0.000000 Rows_sent: 1  Rows_examined: 0
SET timestamp=1585908905;
SELECT sleep(5);
mysql> select @@version;
+------------+
| @@version  |
+------------+
| 5.7.23-log |
+------------+
1 row in set (0.00 sec)

```
### 3. What did you see instead?
**In TiDB :**
<img width="927" alt="企业微信截图_6fb4caaf-5c4e-4dcd-a162-9e06da06417e" src="https://user-images.githubusercontent.com/20888467/78349160-bdbd3900-75d5-11ea-9c3d-cba46db02307.png">

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
v2.1.17



## SIG slack channel
[#sig-exec](https://tidbcommunity.slack.com/archives/CMRD79DRR)

## Score
- 300

## Mentor
 * @qw4990

### PR DESCRIPTION
**Pull request by nayuta-yanagisawa**
### What problem does this PR solve?

Issue Number: close #16062

Problem Summary: With a slow prepared statement, the slow query log records it in the form of `exec ...` rather than the underlying SQL with placeholders. 

### What is changed and how it works?

What's Changed: Records original queries of prepared statements.

How it Works:

On master branch (53a68b56da072c6b9863ce38b4488ede618cfd49)

```
mysql> PREPARE mystmt FROM 'SELECT SLEEP(?)';
Query OK, 0 rows affected (0.00 sec)

mysql> SET @num = 1;
Query OK, 0 rows affected (0.00 sec)

mysql> EXECUTE mystmt USING @num;
+----------+
| SLEEP(?) |
+----------+
|        0 |
+----------+
1 row in set (1.00 sec)

mysql> SELECT Query FROM information_schema.SLOW_QUERY LIMIT 1;
+-------------------------------------------+
| Query                                     |
+-------------------------------------------+
| EXECUTE mystmt USING @num [arguments: 1]; |
+-------------------------------------------+
1 row in set (0.00 sec)
```

On PR branch

```
mysql> PREPARE mystmt FROM 'SELECT SLEEP(?)';
Query OK, 0 rows affected (0.00 sec)

mysql> SET @num = 1;
Query OK, 0 rows affected (0.00 sec)

mysql> EXECUTE mystmt USING @num;
+----------+
| SLEEP(?) |
+----------+
|        0 |
+----------+
1 row in set (1.00 sec)

mysql> SELECT Query FROM information_schema.SLOW_QUERY LIMIT 1;
+---------------------------------+
| Query                           |
+---------------------------------+
| SELECT SLEEP(?) [arguments: 1]; |
+---------------------------------+
1 row in set (0.00 sec)
```

### Related changes

N/A

### Check List 

Tests 

- Integration test

Side effects

- Breaking backward compatibility

### Release note 

- Slow query log records original queries with placeholders for prepared statements.

### CODE REVIEW

**Comment by nayuta-yanagisawa** [comment]
All tests now pass. Could you take a look at the present pull request? @qw4990

File: executor/adapter.go [hunk 1]
```diff
@@ -886,15 +880,8 @@ func (a *ExecStmt) LogSlowQuery(txnTS uint64, succ bool, hasMoreResults bool) {
...
-	var sql stringutil.StringerFunc
-	normalizedSQL, digest := sessVars.StmtCtx.SQLDigest()
-	if sessVars.EnableRedactLog {
-		sql = FormatSQL(normalizedSQL, nil)
-	} else if sensitiveStmt, ok := a.StmtNode.(ast.SensitiveStmtNode); ok {
-		sql = FormatSQL(sensitiveStmt.SecureText(), nil)
-	} else {
-		sql = FormatSQL(a.Text, sessVars.PreparedParams)
-	}
+	sql := FormatSQL(a.GetTextToLog())
```
Comment (xhebox) [inline executor/adapter.go:883]: I am afraid of that the new code broke the original semantic, which is introduced in #18578 by @crazycs520 , PTAL at this PR.

**Comment by qw4990** [comment]
Please resolve conflicts at your convenience @nayuta-yanagisawa

**Comment by nayuta-yanagisawa** [comment]
I've resolved the conflict. @qw4990

**Review by xhebox — APPROVED** [review]
After a deeper look, I think it is correct. LGTM

**Comment by nayuta-yanagisawa** [comment]
The CI fails continuously. I think that this is possibly fixed by https://github.com/pingcap/tidb/issues/23967.

**Pull request merged**